### ISSUE TITLE
"Call method" increase readability of code

### ISSUE THREAD

**Issue by eiseleh** [SAP/styleguides#194]
In my opinion the coding is readable if it is like a natural language. "Call method" is therefore better readable than brackets at the end. If "call method" is not used the reader always have to look at the end of the object which will be used, to see whether a method is called and not a property. 
It is clear that some use cases especially if you are working with returning parameter, does not allow to use call method. But I think it is ok. to use different patterns if you have different use cases.

**Comment by xtough** [comment]
I would claim that if you feel the need to prepend "call method" to a method invocation, you have done something wrong in your API design. If you do proper OO in the form of ``my_object->do_something( ... )`` I personally find this more readable than the "call method" variant. To that end, "call method" is even detrimental to the use of [fluent interfaces](https://en.wikipedia.org/wiki/Fluent_interface).

I think this is also one reason why "call method" is labeled as [obsolete for static calls](https://help.sap.com/doc/abapdocu_752_index_htm/7.52/en-us/abapcall_method_static.htm) and the only recommended use of the "call method" statement is for [dynamic method invocations](https://help.sap.com/doc/abapdocu_752_index_htm/7.52/en-us/abenmethod_calls_dynamic.htm) to clearly distinguish them from static calls.

I would propose to close this issue.

**Comment by eiseleh** [comment]
"Call method" is better readable and clearly signals the the existing procedure will be left. From my point of view this is a really nice language element of ABAP which should not be omitted.  I think it is rather unfair to say that I have done something wrong.

**Comment by xtough** [comment]
No offense meant. It's actually not your fault but the fault of the provider of the API you are using. Let me rephrase it: If your are faced with a class that looks like a function module because contains just static methods, e.g. ``cl_my_utility=>magical_conversion( ... )``, you may feel the need to treat this like a function call and hence see a need to invoke this via ``call method``. But this is actually in indicator that the provider of ``cl_my_utility`` has not fully arrived in the world of object-oriented programming. 
Besides, in object-oriented code there are typically so many indirections to other methods that you would get overwhelmed by too many ``call method`` statements.

**Comment by nununo** [comment]
@eiseleh I wouldn't go as far as saying you're doing something wrong.

But I think the only reason you say it's more readable is because you're used to it, which is not in itself a good enough argument.

Consider the following comparison.

With CALL METHOD:
```ABAP
DATA result TYPE i.
CALL METHOD method1
  IMPORTING
    x = a
  RECEIVING
    y = a.

CALL METHOD method2
  IMPORTING
    x = a
  RECEIVING
    y = a.

CALL METHOD method3
  IMPORTING
    x = a
  RECEIVING
    y = a.

CALL METHOD method4
  IMPORTING
    x = a
  RECEIVING
    y = a.
```

Functional syntax:
```ABAP
DATA(result) = method1( 3 ).
a = method2( a ).
a = method3( a ).
a = method4( a ).
```

And this is yet another possibility:
```ABAP
DATA(a) = method4( method3( method2( method1( 3 ) ) ) ).
```

This last one is rather confusing because in this example we have 4 method calls. But in many cases where just two are needed it is clearly a good approach.

Now I'll give you another example which should make it even more clear that CALL METHOD should be abandoned.

This is with CALL METHOD:
```ABAP
DATA o TYPE zcl_class1.
DATA a TYPE i.
CALL METHOD get_class
  RECEIVING
    class = o.

CALL METHOD o->method1
  IMPORTING
    x = 3
  RECEVING
    result = a.
```

And this is without CALL METHOD:
```ABAP
DATA(a) = get_class( )->method1( 3 ).
```

Using CALL METHOD requires 11 lines. Without  CALL METHOD... only 1! Do you still think it is more readable?

**Comment by eiseleh** [comment]
Yes, I guess basically you are right. But I think in S4 we need to use legacy coding for longer than a decade and we have to cope with it. If the called method cl_my_utility=>magical_conversion( ... ) has some modern substitute we can refactor our coding. This might than be the right time to omit "Call Method".

**Comment by schneiderand** [comment]
Maybe it's sometimes more readable. But in my opinion it's not easier to use while debugging, especially if there are more than 2 methods in a chain or if such chains are even used inline in a parameter position of some "outer" method (all seen several times and hated it).
An besides this, why are we are trying to do code beautification via ATC checks? 
We should leave this tool for finding real errors.

**Comment by xtough** [comment]
> An besides this, why are we are trying to do code beautification via ATC checks?

Technically speaking, this question does not belong in this project but to https://github.com/SAP/code-pal-for-abap/ where the ATC checks for clean ABAP are implemented.

> We should leave this tool for finding real errors.

That's a personal opinion you are of course entitled to but the decision which issues are treated with which severity is up to your system administrators. You should challenge them with your concern.

Besides, there are good reasons why code style does matter and can help reducing defects. Code usually spends much more time in maintenance than in development and is read by way more people than it was originally written by. Hence it pays off to spend more time on optimizing it for readability so that reviewers can analyze quicker.

**Comment by schneiderand** [comment]
Well, according to my experience, debugging of code is something which happens very often during maintenance.
And don't you think that a bad "writability" of code doesn't increase the risk of bugs?

**Comment by pokrakam** [comment]
Another more subtle aspect is ABAP's evolution from a pure procedural language into a more [declarative](https://en.wikipedia.org/wiki/Declarative_programming) paradigm. The `CALL METHOD` format is an imperative statement that firmly belongs in the procedural world. 

The functional format is more versatile, CALL METHOD must always be at the beginning of a statement. Only the function syntax can be used in an operand position. So it makes sense to harmonise on the one style that is both more compact and more versatile and limit the more restrictive notation only to where necessary. 

Although would make logical sense to do `o->(x)( ).` or `(objname)->(methodname)( )`

Regarding brackets, this function/method invocation syntax `func( )` is also widely used in many other programming languages, so this is another angle to consider.

**Comment by eiseleh** [comment]
I don't say that Call Method has to be set for each method call. In a completely new ABAP coding with no relation to old objects the young developer with java background will do method chaining and may don't miss to use call method to emphasis the call of a big method of another area. 
However, we will not have the time in S4 to get rid of all existing things. And Call Method gives the chance to emphasis the call of another class which must probably does have a procedural style.  With that you easily find it in the editor by the syntax highlighting. If you omit that you only have the chance do put attention to the call by comments.
Regarding readability I still stick to my opinion that it will improve as more as it is a natural language: If you describe the call of a method face to face to a colleague you would say. The method x of class C1 **calls method** of class C1.  Therefore ABAP allows to write more readable code as in many other languages.
In contrast to eslint the new style is now forced to applied by ATC for all use cases. But it really does not fit for all use cases.

**Comment by xtough** [comment]
I see your point. But again, this is something you need to clarify with your system administrators who are responsible for configuring ATC priorities. 

The Clean ABAP Style Guide focusses on how to write modern ABAP code. It appears to me that the opinion of the community's majority as well as of the official documentation in that regard is *NOT* in favor of ``call method``.

**Comment by schneiderand** [comment]
"Real world" ABAP example taken from a current implementation (not written by me). Where is the method call?
Don't you think a CALL METHOD  statement instead would catch your eye much faster?

  METHOD prep_general_field_prop.

    IF  is_defect-defectlifecyclestatus EQ if_i_defect_status_util=>cs_status_value-defectlifecyclestatus-completed OR
        is_defect-defectlifecyclestatus EQ if_i_defect_status_util=>cs_status_value-defectlifecyclestatus-not_relevant OR
        is_defect-isdeleted EQ abap_true.
      cs_properties-prop_read_only_new = abap_true.
      cs_properties-r_changed          = abap_true.
    ELSEIF is_defect-defectlifecyclestatus EQ if_i_defect_status_util=>cs_status_value-defectlifecyclestatus-in_process.
      cs_properties-prop_read_only_new = abap_false.
      cs_properties-r_changed          = abap_true.
    ENDIF.

    IF cl_i_defect_draft_util=>get_instance( )->get_defecthasaffectedobjects( is_defect-defectcategory ) EQ abap_true.
      IF ( v_fieldname EQ if_i_defect_tp_c=>sc_node_attribute-i_defect_tp-defectivequantity OR
           iv_fieldname EQ if_i_defect_tp_c=>sc_node_attribute-i_defect_tp-defectivequantityunit ).
        cs_properties-prop_read_only_new = abap_true.
        cs_properties-r_changed          = abap_true.
      ENDIF.
      [...]

    ENDIF.
 [...]

**Comment by xtough** [comment]
As @nununo wrote earlier in this thread:
> But I think the only reason you say it's more readable is because you're used to it, which is not in itself a good enough argument.

You may also tweak your editor's syntax highlighting. For instance, in ABAP in Eclipse (ADT) special characters like ``( -> => )`` etc. are highlighted in blue, which makes it easier to spot method calls than with the standard SE80 settings.

**Comment by HrFlorianHoffmann** [comment]
@schneiderand There is a variety of possibilities to improve the readability of the method you quoted.

The complete method looks like it would benefit from splitting. It has 30 statements on 65 lines, which in itself is already quite long. The section [Keep methods small](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#keep-methods-small) suggests to reduce that size.

The method is segmented into four blocks that are separated by blank lines. According to the section [Add a single blank line to separate things, but not more](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#add-a-single-blank-line-to-separate-things-but-not-more), this visual pattern is an indicator that the method may do more than one thing. The section [Do one thing, do it well, do it only](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#do-one-thing-do-it-well-do-it-only) suggests we should probably split this method.

The four blocks start with conditions of the form `IF iv_fieldname = {some_constant}` that exclude each other. This suggests that it subsumes four independent processing steps that could be moved out into independent sub-methods. 

The three comments at the start of some of the blocks are typical cases of comments that attempt to fix missing method names. The section [Use methods instead of comments to segment your code](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#use-methods-instead-of-comments-to-segment-your-code) suggests that the blocks should be moved into sub-methods whose names reveal those intentions.

There is some duplication in that method, in the form of a mini block that appears several times:
    
    cs_properites-prop_read_only_new = {abap_true/abap_false}.
    cs_properties-r_changed = abap_true
    
Duplication should be avoided, as per the _Don't Repeat Yourself_ principle. This duplication could be removed by extracting a utility method `set_read_only_property`. The section [Split method instead of Boolean input parameter](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#split-method-instead-of-boolean-input-parameter) suggests that two utility methods `make_read_only` and `make_write_enabled` might be even better:

    METHOD make_read_only.
      cs_properites-prop_read_only_new = abap_true.
      cs_properties-r_changed = abap_true.     
    ENDMETHOD.

The complexity of the conditions suggests that they could benefit from extraction into sub-methods whose names reveal their intention, as per [Consider extracting complex conditions](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#consider-extracting-complex-conditions). This method seems to be at the heart of business logic, but it does not really reveal that business logic in an obvious way.
 
There is also some duplication in the conditions that you could remove by doing so. For example, the comparison whether `defectlifecyclestatus` is `completed` or `not_relevant` appears in two distinct places. Chances are that you will forget to update one if you change the other. Even more important, that exact same combination of conditions appears multiple times in other classes as well, suggesting that it should be moved to a more central place in your application code.

The long line of code that you are specifically referring to, with the in-lined chained method calls,

    IF cl_i_defect_draft_util=>get_instance( )->get_defecthasaffectedobjects( is_defect-defectcategory ) EQ abap_true.
    
has been turned into a comment, and replaced with a variable lookup that looks already much more agreeable:

    IF <ls_defectcatergory_data>-defecthasaffectedobjects EQ abap_true.
    
Doing this kind of refactoring is what Clean ABAP suggests in [Consider decomposing complex conditions](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#consider-decomposing-complex-conditions).

You should also consider removing that commented line to reduce visual clutter, as per [Delete code instead of commenting it](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#delete-code-instead-of-commenting-it).

The [ABAP Language Help article _rel_exp - Binary Relational Operators_](https://help.sap.com/doc/abapdocu_752_index_htm/7.52/en-us/abenlogexp_any_operand.htm) suggests to prefer the simple equality comparator `=` to `EQ` in this place, because "if in doubt, the variant with the characters =, <, and > is considered to be more up-to-date", whereas the two-letter comparators "are better suited to other relational operators such as CO, CN, and so on".

This is a superficial review based on the structure of the code, meaning I don't really understand what your code does, and what's around it. Diving into the semantics and context may reveal deeper potential for changes; that's why [we encourage coders to perform code reviews](https://github.com/SAP/styleguides/tree/main/abap-code-review).

My impression is that by trying out some of these suggestions you will find a form for that method that the team will find more than agreeable. None of these changes requires the `CALL METHOD` statement. I agree with @xtough on that point: If I researched correctly, that statement was marked as obsolete in SAP NetWeaver 7.40 in 2014. That was 7 years ago. Even if you disagree with Clean ABAP's suggestion to [Avoid obsolete language elements](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#avoid-obsolete-language-elements), we should also consider [the ABAP language group's recommendation](https://help.sap.com/doc/abapdocu_740_index_htm/7.40/en-us/abenabap_obsolete.htm) that "[t]hese statements may still be encountered in older programs but should not be used in new programs."

As you and @eiseleh point out, that does not necessarily mean that we should replace `CALL METHOD` everywhere and immediately. Me personally, I am more a friend of [the _boy scout rule_](https://github.com/SAP/styleguides/blob/main/clean-abap/CleanABAP.md#how-to-refactor-legacy-code) that promotes small, iterative improvements over longer periods of time. I also agree with @xtough on the point that the work mode you choose for refactoring is not exactly part of Clean ABAP - you should discuss it with your team, and/or the larger team around it.

A quick coverage measurement in one of the sandbox systems by the way suggests that this method has excellent unit test coverage, with 94.11 % of branches covered, missing only 1 out of 17. Without confirming the quality of these unit tests, this is an impressive number and sounds like a solid basis for trying out refactoring options without constant fear of breaking something.

**Comment by xtough** [comment]
Can we close this issue? Feedback seems to be stagnant for 2 weeks now.

**Comment by eiseleh** [comment]
Thanks for the feedback. Issue can be closed.
